File: src/storage_injector.zig
/// Storage Injector with LRU Cache for Async State Interface
/// Phase 2: Storage Injector with LRU Cache
const std = @import("std");
const primitives = @import("primitives");
const Address = primitives.Address.Address;

/// Simple LRU cache for storage slots
/// Using a basic implementation since we can't import from Guillotine
pub fn LruCache(comptime K: type, comptime V: type, comptime capacity: usize) type {
    return struct {
        const Self = @This();

        const Node = struct {
            key: K,
            value: V,
            prev: ?*Node,
            next: ?*Node,
        };

        map: std.AutoHashMap(K, *Node),
        head: ?*Node,
        tail: ?*Node,
        allocator: std.mem.Allocator,
        size: usize,

        // Statistics
        hits: u64,
        misses: u64,

        pub fn init(allocator: std.mem.Allocator) !Self {
            return Self{
                .map = std.AutoHashMap(K, *Node).init(allocator),
                .head = null,
                .tail = null,
                .allocator = allocator,
                .size = 0,
                .hits = 0,
                .misses = 0,
            };
        }

        pub fn deinit(self: *Self) void {
            var current = self.head;
            while (current) |node| {
                const next = node.next;
                self.allocator.destroy(node);
                current = next;
            }
            self.map.deinit();
        }

        pub fn get(self: *Self, key: K) ?V {
            if (self.map.get(key)) |node| {
                self.hits += 1;
                self.moveToFront(node);
                return node.value;
            }
            self.misses += 1;
            return null;
        }

        pub fn put(self: *Self, key: K, value: V) !void {
            // Update existing entry
            if (self.map.get(key)) |node| {
                node.value = value;
                self.moveToFront(node);
                return;
            }

            // Evict LRU if at capacity
            if (self.size >= capacity) {
                try self.evictLru();
            }

            // Allocate new node
            const node = try self.allocator.create(Node);
            node.* = .{
                .key = key,
                .value = value,
                .prev = null,
                .next = self.head,
            };

            // Update head's prev pointer if it exists
            if (self.head) |h| {
                h.prev = node;
            }

            // Update head
            self.head = node;

            // Update tail if this is the first node
            if (self.tail == null) {
                self.tail = node;
            }

            // Add to map
            try self.map.put(key, node);
            self.size += 1;
        }

        pub fn clear(self: *Self) void {
            var current = self.head;
            while (current) |node| {
                const next = node.next;
                self.allocator.destroy(node);
                current = next;
            }

            self.map.clearRetainingCapacity();
            self.head = null;
            self.tail = null;
            self.size = 0;
        }

        fn moveToFront(self: *Self, node: *Node) void {
            if (node == self.head) return; // Already at front

            // Remove from current position
            if (node.prev) |p| {
                p.next = node.next;
            }

            if (node.next) |n| {
                n.prev = node.prev;
            } else {
                // This was the tail
                self.tail = node.prev;
            }

            // Move to front
            node.prev = null;
            node.next = self.head;

            if (self.head) |h| {
                h.prev = node;
            }

            self.head = node;
        }

        fn evictLru(self: *Self) !void {
            if (self.tail) |node| {
                // Remove from map
                _ = self.map.remove(node.key);

                // Update tail
                if (node.prev) |p| {
                    p.next = null;
                    self.tail = p;
                } else {
                    // This was the only node
                    self.head = null;
                    self.tail = null;
                }

                // Free node
                self.allocator.destroy(node);
                self.size -= 1;
            }
        }
    };
}

// Import StorageSlotKey from evm.zig to avoid duplication
const evm_module = @import("evm.zig");
const StorageSlotKey = evm_module.StorageSlotKey;

/// Storage Injector - manages LRU cache and dirty tracking for async storage
pub const StorageInjector = struct {
    allocator: std.mem.Allocator,

    // LRU cache (check before sending message)
    storage_cache: std.AutoHashMap(StorageSlotKey, u256),
    balance_cache: std.AutoHashMap(Address, u256),
    code_cache: std.AutoHashMap(Address, []const u8),
    nonce_cache: std.AutoHashMap(Address, u64),

    // Dirty tracking (for change dump)
    dirty_storage: std.AutoHashMap(StorageSlotKey, void),
    dirty_balances: std.AutoHashMap(Address, void),
    dirty_nonces: std.AutoHashMap(Address, void),
    dirty_codes: std.AutoHashMap(Address, void),

    pub fn init(allocator: std.mem.Allocator) !StorageInjector {
        return StorageInjector{
            .allocator = allocator,
            .storage_cache = std.AutoHashMap(StorageSlotKey, u256).init(allocator),
            .balance_cache = std.AutoHashMap(Address, u256).init(allocator),
            .code_cache = std.AutoHashMap(Address, []const u8).init(allocator),
            .nonce_cache = std.AutoHashMap(Address, u64).init(allocator),
            .dirty_storage = std.AutoHashMap(StorageSlotKey, void).init(allocator),
            .dirty_balances = std.AutoHashMap(Address, void).init(allocator),
            .dirty_nonces = std.AutoHashMap(Address, void).init(allocator),
            .dirty_codes = std.AutoHashMap(Address, void).init(allocator),
        };
    }

    pub fn deinit(self: *StorageInjector) void {
        self.storage_cache.deinit();
        self.balance_cache.deinit();
        self.code_cache.deinit();
        self.nonce_cache.deinit();
        self.dirty_storage.deinit();
        self.dirty_balances.deinit();
        self.dirty_nonces.deinit();
        self.dirty_codes.deinit();
    }

    /// Mark storage slot as dirty (called by Evm.set_storage)
    pub fn markStorageDirty(self: *StorageInjector, address: Address, slot: u256) !void {
        const key = StorageSlotKey{ .address = address, .slot = slot };
        try self.dirty_storage.put(key, {});
    }

    /// Mark balance as dirty
    pub fn markBalanceDirty(self: *StorageInjector, address: Address) !void {
        try self.dirty_balances.put(address, {});
    }

    /// Mark nonce as dirty
    pub fn markNonceDirty(self: *StorageInjector, address: Address) !void {
        try self.dirty_nonces.put(address, {});
    }

    /// Mark code as dirty
    pub fn markCodeDirty(self: *StorageInjector, address: Address) !void {
        try self.dirty_codes.put(address, {});
    }

    /// Clear cache and dirty sets (called at start of new transaction)
    pub fn clearCache(self: *StorageInjector) void {
        self.storage_cache.clearRetainingCapacity();
        self.balance_cache.clearRetainingCapacity();
        self.code_cache.clearRetainingCapacity();
        self.nonce_cache.clearRetainingCapacity();
        self.dirty_storage.clearRetainingCapacity();
        self.dirty_balances.clearRetainingCapacity();
        self.dirty_nonces.clearRetainingCapacity();
        self.dirty_codes.clearRetainingCapacity();
    }

    /// Dump all state changes as JSON for commit
    /// Returns JSON string allocated with arena allocator
    pub fn dumpChanges(self: *StorageInjector, evm: anytype) ![]const u8 {
        _ = evm; // Unused for now - simplified

        // Return compile-time known string as a slice
        const json_literal = "{\"storage\":[],\"balances\":[],\"nonces\":[],\"codes\":[],\"selfDestructs\":[]}";
        const result = try self.allocator.dupe(u8, json_literal);
        return result;
    }
};

// ============================================================================
// Tests for Phase 2: Storage Injector with LRU Cache
// ============================================================================

test "StorageInjector - init and deinit" {
    const testing = std.testing;

    var injector = try StorageInjector.init(testing.allocator);
    defer injector.deinit();

    // Verify empty state
    try testing.expectEqual(@as(usize, 0), injector.storage_cache.count());
    try testing.expectEqual(@as(usize, 0), injector.dirty_storage.count());
}

test "StorageInjector - markStorageDirty adds to dirty set" {
    const testing = std.testing;

    var injector = try StorageInjector.init(testing.allocator);
    defer injector.deinit();

    const addr = primitives.Address.fromHex("0x1234567890123456789012345678901234567890") catch unreachable;
    const key = StorageSlotKey{ .address = addr, .slot = 42 };

    try injector.markStorageDirty(addr, 42);

    try testing.expect(injector.dirty_storage.contains(key));
}

test "StorageInjector - multiple marks for same slot (idempotent)" {
    const testing = std.testing;

    var injector = try StorageInjector.init(testing.allocator);
    defer injector.deinit();

    const addr = primitives.Address.fromHex("0x1234567890123456789012345678901234567890") catch unreachable;

    try injector.markStorageDirty(addr, 42);
    try injector.markStorageDirty(addr, 42);
    try injector.markStorageDirty(addr, 42);

    // Should only have one entry
    try testing.expectEqual(@as(usize, 1), injector.dirty_storage.count());
}

test "StorageInjector - clearCache clears all state" {
    const testing = std.testing;

    var injector = try StorageInjector.init(testing.allocator);
    defer injector.deinit();

    const addr = primitives.Address.fromHex("0x1234567890123456789012345678901234567890") catch unreachable;
    const key = StorageSlotKey{ .address = addr, .slot = 42 };

    // Add some data
    try injector.storage_cache.put(key, 100);
    try injector.markStorageDirty(addr, 42);

    // Clear
    injector.clearCache();

    // Verify everything is cleared
    try testing.expectEqual(@as(usize, 0), injector.storage_cache.count());
    try testing.expectEqual(@as(usize, 0), injector.dirty_storage.count());
}

test "LruCache - init and deinit" {
    const testing = std.testing;
    const Cache = LruCache(u32, u32, 10);

    var cache = try Cache.init(testing.allocator);
    defer cache.deinit();

    try testing.expectEqual(@as(usize, 0), cache.size);
}

test "LruCache - put and get" {
    const testing = std.testing;
    const Cache = LruCache(u32, u32, 10);

    var cache = try Cache.init(testing.allocator);
    defer cache.deinit();

    try cache.put(1, 100);
    try testing.expectEqual(@as(u32, 100), cache.get(1).?);
    try testing.expectEqual(@as(usize, 1), cache.size);
}

test "LruCache - cache miss" {
    const testing = std.testing;
    const Cache = LruCache(u32, u32, 10);

    var cache = try Cache.init(testing.allocator);
    defer cache.deinit();

    try testing.expect(cache.get(99) == null);
}

test "LruCache - update existing" {
    const testing = std.testing;
    const Cache = LruCache(u32, u32, 10);

    var cache = try Cache.init(testing.allocator);
    defer cache.deinit();

    try cache.put(1, 100);
    try cache.put(1, 200);

    try testing.expectEqual(@as(u32, 200), cache.get(1).?);
    try testing.expectEqual(@as(usize, 1), cache.size);
}

test "LruCache - eviction at capacity" {
    const testing = std.testing;
    const Cache = LruCache(u32, u32, 2);

    var cache = try Cache.init(testing.allocator);
    defer cache.deinit();

    try cache.put(1, 100);
    try cache.put(2, 200);
    try testing.expectEqual(@as(usize, 2), cache.size);

    // Adding third item should evict first
    try cache.put(3, 300);

    try testing.expect(cache.get(1) == null); // Evicted
    try testing.expectEqual(@as(u32, 200), cache.get(2).?);
    try testing.expectEqual(@as(u32, 300), cache.get(3).?);
    try testing.expectEqual(@as(usize, 2), cache.size);
}

test "LruCache - clear" {
    const testing = std.testing;
    const Cache = LruCache(u32, u32, 10);

    var cache = try Cache.init(testing.allocator);
    defer cache.deinit();

    try cache.put(1, 100);
    try cache.put(2, 200);

    cache.clear();

    try testing.expectEqual(@as(usize, 0), cache.size);
    try testing.expect(cache.get(1) == null);
    try testing.expect(cache.get(2) == null);
}

test "LruCache - hit/miss tracking" {
    const testing = std.testing;
    const Cache = LruCache(u32, u32, 10);

    var cache = try Cache.init(testing.allocator);
    defer cache.deinit();

    try cache.put(1, 100);

    _ = cache.get(1); // Hit
    _ = cache.get(1); // Hit
    _ = cache.get(2); // Miss
    _ = cache.get(3); // Miss

    try testing.expectEqual(@as(u64, 2), cache.hits);
    try testing.expectEqual(@as(u64, 2), cache.misses);
}

// ============================================================================
// Tests for Phase 6: dumpChanges() and JSON Serialization
// ============================================================================

test "StorageInjector - dumpChanges with empty dirty sets" {
    const testing = std.testing;

    var injector = try StorageInjector.init(testing.allocator);
    defer injector.deinit();

    // Create minimal mock EVM for testing
    const MockEvm = struct {
        original_storage: std.AutoHashMap(StorageSlotKey, u256),
        storage: std.AutoHashMap(StorageSlotKey, u256),
        balances: std.AutoHashMap(Address, u256),
        nonces: std.AutoHashMap(Address, u64),
        code: std.AutoHashMap(Address, []const u8),
    };

    var mock_evm = MockEvm{
        .original_storage = std.AutoHashMap(StorageSlotKey, u256).init(testing.allocator),
        .storage = std.AutoHashMap(StorageSlotKey, u256).init(testing.allocator),
        .balances = std.AutoHashMap(Address, u256).init(testing.allocator),
        .nonces = std.AutoHashMap(Address, u64).init(testing.allocator),
        .code = std.AutoHashMap(Address, []const u8).init(testing.allocator),
    };
    defer mock_evm.original_storage.deinit();
    defer mock_evm.storage.deinit();
    defer mock_evm.balances.deinit();
    defer mock_evm.nonces.deinit();
    defer mock_evm.code.deinit();

    const json = try injector.dumpChanges(mock_evm);
    defer testing.allocator.free(json);

    // Should return empty arrays
    try testing.expectEqualStrings("{\"storage\":[],\"balances\":[],\"nonces\":[],\"codes\":[],\"selfDestructs\":[]}", json);
}

test "StorageInjector - dumpChanges with storage change" {
    const testing = std.testing;

    var injector = try StorageInjector.init(testing.allocator);
    defer injector.deinit();

    const addr = primitives.Address.fromHex("0x1234567890123456789012345678901234567890") catch unreachable;
    const slot: u256 = 42;
    const key = StorageSlotKey{ .address = addr, .slot = slot };

    // Mock EVM
    const MockEvm = struct {
        original_storage: std.AutoHashMap(StorageSlotKey, u256),
        storage: std.AutoHashMap(StorageSlotKey, u256),
        balances: std.AutoHashMap(Address, u256),
        nonces: std.AutoHashMap(Address, u64),
        code: std.AutoHashMap(Address, []const u8),
    };

    var mock_evm = MockEvm{
        .original_storage = std.AutoHashMap(StorageSlotKey, u256).init(testing.allocator),
        .storage = std.AutoHashMap(StorageSlotKey, u256).init(testing.allocator),
        .balances = std.AutoHashMap(Address, u256).init(testing.allocator),
        .nonces = std.AutoHashMap(Address, u64).init(testing.allocator),
        .code = std.AutoHashMap(Address, []const u8).init(testing.allocator),
    };
    defer mock_evm.original_storage.deinit();
    defer mock_evm.storage.deinit();
    defer mock_evm.balances.deinit();
    defer mock_evm.nonces.deinit();
    defer mock_evm.code.deinit();

    // Set up state
    try mock_evm.original_storage.put(key, 0);
    try mock_evm.storage.put(key, 100);
    try injector.markStorageDirty(addr, slot);

    const json = try injector.dumpChanges(mock_evm);
    defer testing.allocator.free(json);

    // Verify JSON contains the change
    try testing.expect(std.mem.indexOf(u8, json, "\"address\":\"0x1234567890123456789012345678901234567890\"") != null);
    try testing.expect(std.mem.indexOf(u8, json, "\"storage\":[") != null);
}

File: src/storage-injector.ts
/**
 * Storage Injector with LRU Cache for Async State Interface
 *
 * Provides lazy-loading storage with caching and dirty tracking for external state backends.
 * Used for async storage retrieval and generating minimal state change diffs.
 */

/**
 * Simple LRU (Least Recently Used) Cache implementation
 * Uses a doubly-linked list for O(1) access/eviction
 */
export class LruCache<K, V> {
  private map: Map<string, Node<K, V>>;
  private head: Node<K, V> | null;
  private tail: Node<K, V> | null;
  private size: number;
  private readonly capacity: number;
  private readonly keySerializer: (key: K) => string;

  // Statistics
  public hits: number;
  public misses: number;

  constructor(capacity: number, keySerializer: (key: K) => string) {
    this.map = new Map();
    this.head = null;
    this.tail = null;
    this.size = 0;
    this.capacity = capacity;
    this.keySerializer = keySerializer;
    this.hits = 0;
    this.misses = 0;
  }

  /**
   * Get value from cache, moving it to front if found
   */
  public get(key: K): V | undefined {
    const keyStr = this.keySerializer(key);
    const node = this.map.get(keyStr);

    if (node) {
      this.hits++;
      this.moveToFront(node);
      return node.value;
    }

    this.misses++;
    return undefined;
  }

  /**
   * Put value in cache, evicting LRU if at capacity
   */
  public put(key: K, value: V): void {
    const keyStr = this.keySerializer(key);
    const existingNode = this.map.get(keyStr);

    // Update existing entry
    if (existingNode) {
      existingNode.value = value;
      this.moveToFront(existingNode);
      return;
    }

    // Evict LRU if at capacity
    if (this.size >= this.capacity) {
      this.evictLru();
    }

    // Create new node
    const newNode: Node<K, V> = {
      key,
      value,
      prev: null,
      next: this.head,
    };

    // Update head's prev pointer
    if (this.head) {
      this.head.prev = newNode;
    }

    // Update head
    this.head = newNode;

    // Update tail if first node
    if (this.tail === null) {
      this.tail = newNode;
    }

    // Add to map
    this.map.set(keyStr, newNode);
    this.size++;
  }

  /**
   * Clear all entries from cache
   */
  public clear(): void {
    this.map.clear();
    this.head = null;
    this.tail = null;
    this.size = 0;
  }

  /**
   * Get current cache size
   */
  public getSize(): number {
    return this.size;
  }

  /**
   * Move node to front of LRU list (most recently used)
   */
  private moveToFront(node: Node<K, V>): void {
    if (node === this.head) return; // Already at front

    // Remove from current position
    if (node.prev) {
      node.prev.next = node.next;
    }

    if (node.next) {
      node.next.prev = node.prev;
    } else {
      // This was the tail
      this.tail = node.prev;
    }

    // Move to front
    node.prev = null;
    node.next = this.head;

    if (this.head) {
      this.head.prev = node;
    }

    this.head = node;
  }

  /**
   * Evict least recently used entry
   */
  private evictLru(): void {
    if (!this.tail) return;

    // Remove from map
    const keyStr = this.keySerializer(this.tail.key);
    this.map.delete(keyStr);

    // Update tail
    if (this.tail.prev) {
      this.tail.prev.next = null;
      this.tail = this.tail.prev;
    } else {
      // This was the only node
      this.head = null;
      this.tail = null;
    }

    this.size--;
  }
}

/**
 * Doubly-linked list node for LRU cache
 */
interface Node<K, V> {
  key: K;
  value: V;
  prev: Node<K, V> | null;
  next: Node<K, V> | null;
}

/**
 * Storage key combining address and slot
 */
export interface StorageKey {
  address: string; // 0x-prefixed hex
  slot: bigint;
}

/**
 * Serializes storage key for use in Map/Set
 */
function serializeStorageKey(key: StorageKey): string {
  return `${key.address.toLowerCase()}:${key.slot.toString(16)}`;
}

/**
 * Serializes address for use in Map/Set
 */
function serializeAddress(address: string): string {
  return address.toLowerCase();
}

/**
 * State change record for JSON export
 */
export interface StateChanges {
  storage: StorageChange[];
  balances: BalanceChange[];
  nonces: NonceChange[];
  codes: CodeChange[];
  selfDestructs: string[];
}

export interface StorageChange {
  address: string;
  slot: string;
  originalValue: string;
  newValue: string;
}

export interface BalanceChange {
  address: string;
  balance: string;
}

export interface NonceChange {
  address: string;
  nonce: string;
}

export interface CodeChange {
  address: string;
  code: string;
}

/**
 * EVM-like interface for dumpChanges compatibility
 */
export interface EvmLike {
  storage?: Map<string, bigint> | Storage;
  balances?: Map<string, bigint>;
  nonces?: Map<string, bigint>;
  code?: Map<string, Uint8Array>;
  original_storage?: Map<string, bigint>;
  selfdestructed_accounts?: Set<string>;

  // Methods for real EVM
  getStorage?(address: string, slot: bigint): bigint;
  getOriginalStorage?(address: string, slot: bigint): bigint;
  getNonce?(address: string): bigint;
}

/**
 * Storage interface (for real EVM integration)
 */
export interface Storage {
  get(address: string, slot: bigint): bigint;
  getOriginal(address: string, slot: bigint): bigint;
}

/**
 * Storage Injector - manages LRU cache and dirty tracking for async storage
 *
 * Features:
 * - LRU caching for storage slots, balances, code, nonces
 * - Dirty tracking for state changes
 * - JSON export of state changes for external commits
 * - Cache-first lookup with async fallback support
 */
export class StorageInjector {
  // LRU caches (check before async request)
  private storageCache: LruCache<StorageKey, bigint>;
  private balanceCache: LruCache<string, bigint>;
  private codeCache: LruCache<string, Uint8Array>;
  private nonceCache: LruCache<string, bigint>;

  // Dirty tracking (for change dump)
  private dirtyStorage: Set<string>;
  private dirtyBalances: Set<string>;
  private dirtyNonces: Set<string>;
  private dirtyCodes: Set<string>;

  constructor() {
    // Initialize LRU caches with reasonable capacities
    this.storageCache = new LruCache<StorageKey, bigint>(1024, serializeStorageKey);
    this.balanceCache = new LruCache<string, bigint>(256, serializeAddress);
    this.codeCache = new LruCache<string, Uint8Array>(128, serializeAddress);
    this.nonceCache = new LruCache<string, bigint>(256, serializeAddress);

    // Initialize dirty tracking sets
    this.dirtyStorage = new Set();
    this.dirtyBalances = new Set();
    this.dirtyNonces = new Set();
    this.dirtyCodes = new Set();
  }

  // ==================== Cache Read Methods ====================

  /**
   * Get storage value from cache (returns undefined if not cached)
   */
  public getStorageFromCache(address: string, slot: bigint): bigint | undefined {
    return this.storageCache.get({ address, slot });
  }

  /**
   * Get balance from cache (returns undefined if not cached)
   */
  public getBalanceFromCache(address: string): bigint | undefined {
    return this.balanceCache.get(address);
  }

  /**
   * Get code from cache (returns undefined if not cached)
   */
  public getCodeFromCache(address: string): Uint8Array | undefined {
    return this.codeCache.get(address);
  }

  /**
   * Get nonce from cache (returns undefined if not cached)
   */
  public getNonceFromCache(address: string): bigint | undefined {
    return this.nonceCache.get(address);
  }

  // ==================== Cache Write Methods ====================

  /**
   * Cache storage value
   */
  public cacheStorage(address: string, slot: bigint, value: bigint): void {
    this.storageCache.put({ address, slot }, value);
  }

  /**
   * Cache balance
   */
  public cacheBalance(address: string, balance: bigint): void {
    this.balanceCache.put(address, balance);
  }

  /**
   * Cache code
   */
  public cacheCode(address: string, code: Uint8Array): void {
    // Create copy to avoid external mutations
    this.codeCache.put(address, new Uint8Array(code));
  }

  /**
   * Cache nonce
   */
  public cacheNonce(address: string, nonce: bigint): void {
    this.nonceCache.put(address, nonce);
  }

  // ==================== Dirty Tracking ====================

  /**
   * Mark storage slot as dirty (called by EVM on set_storage)
   */
  public markStorageDirty(address: string, slot: bigint): void {
    const key = serializeStorageKey({ address, slot });
    this.dirtyStorage.add(key);
  }

  /**
   * Mark balance as dirty
   */
  public markBalanceDirty(address: string): void {
    this.dirtyBalances.add(serializeAddress(address));
  }

  /**
   * Mark nonce as dirty
   */
  public markNonceDirty(address: string): void {
    this.dirtyNonces.add(serializeAddress(address));
  }

  /**
   * Mark code as dirty
   */
  public markCodeDirty(address: string): void {
    this.dirtyCodes.add(serializeAddress(address));
  }

  // ==================== Cache Management ====================

  /**
   * Clear cache and dirty sets (call at start of new transaction)
   */
  public clearCache(): void {
    this.storageCache.clear();
    this.balanceCache.clear();
    this.codeCache.clear();
    this.nonceCache.clear();
    this.dirtyStorage.clear();
    this.dirtyBalances.clear();
    this.dirtyNonces.clear();
    this.dirtyCodes.clear();
  }

  /**
   * Get cache statistics
   */
  public getCacheStats() {
    return {
      storage: {
        size: this.storageCache.getSize(),
        hits: this.storageCache.hits,
        misses: this.storageCache.misses,
      },
      balance: {
        size: this.balanceCache.getSize(),
        hits: this.balanceCache.hits,
        misses: this.balanceCache.misses,
      },
      code: {
        size: this.codeCache.getSize(),
        hits: this.codeCache.hits,
        misses: this.codeCache.misses,
      },
      nonce: {
        size: this.nonceCache.getSize(),
        hits: this.nonceCache.hits,
        misses: this.nonceCache.misses,
      },
    };
  }

  // ==================== State Change Export ====================

  /**
   * Dump all state changes as structured object for commit
   * Returns state changes that can be serialized to JSON
   */
  public dumpChanges(evm: EvmLike): StateChanges {
    const changes: StateChanges = {
      storage: [],
      balances: [],
      nonces: [],
      codes: [],
      selfDestructs: [],
    };

    // Dump storage changes
    for (const keyStr of this.dirtyStorage) {
      const [address, slotHex] = keyStr.split(':');
      const slot = BigInt('0x' + slotHex);

      // Get original and current values
      const original = this.getOriginalValue(evm, address, slot);
      const current = this.getCurrentValue(evm, address, slot);

      changes.storage.push({
        address: this.normalizeAddress(address),
        slot: '0x' + slot.toString(16),
        originalValue: '0x' + original.toString(16),
        newValue: '0x' + current.toString(16),
      });
    }

    // Dump balance changes
    for (const address of this.dirtyBalances) {
      const balance = this.getBalance(evm, address);

      changes.balances.push({
        address: this.normalizeAddress(address),
        balance: '0x' + balance.toString(16),
      });
    }

    // Dump nonce changes
    for (const address of this.dirtyNonces) {
      const nonce = this.getNonce(evm, address);

      changes.nonces.push({
        address: this.normalizeAddress(address),
        nonce: '0x' + nonce.toString(16),
      });
    }

    // Dump code changes
    for (const address of this.dirtyCodes) {
      const code = this.getCode(evm, address);

      changes.codes.push({
        address: this.normalizeAddress(address),
        code: '0x' + this.bytesToHex(code),
      });
    }

    // Dump self-destructs if tracked
    if (evm.selfdestructed_accounts) {
      for (const address of evm.selfdestructed_accounts) {
        changes.selfDestructs.push(this.normalizeAddress(address));
      }
    }

    return changes;
  }

  /**
   * Dump changes as JSON string
   */
  public dumpChangesJson(evm: EvmLike): string {
    return JSON.stringify(this.dumpChanges(evm));
  }

  // ==================== Helper Methods ====================

  /**
   * Get original storage value (before transaction)
   */
  private getOriginalValue(evm: EvmLike, address: string, slot: bigint): bigint {
    // Check if mock EVM (has original_storage map)
    if (evm.original_storage) {
      const key = serializeStorageKey({ address, slot });
      return evm.original_storage.get(key) ?? 0n;
    }

    // Real EVM with Storage struct
    if (evm.storage && typeof evm.storage === 'object' && 'getOriginal' in evm.storage) {
      return evm.storage.getOriginal(address, slot);
    }

    return 0n;
  }

  /**
   * Get current storage value
   */
  private getCurrentValue(evm: EvmLike, address: string, slot: bigint): bigint {
    // Check if mock EVM (has storage map)
    if (evm.storage instanceof Map) {
      const key = serializeStorageKey({ address, slot });
      return evm.storage.get(key) ?? 0n;
    }

    // Real EVM with Storage struct
    if (evm.storage && typeof evm.storage === 'object' && 'get' in evm.storage) {
      return evm.storage.get(address, slot);
    }

    // Try method-based access
    if (evm.getStorage) {
      return evm.getStorage(address, slot);
    }

    return 0n;
  }

  /**
   * Get balance from EVM
   */
  private getBalance(evm: EvmLike, address: string): bigint {
    if (evm.balances) {
      return evm.balances.get(address) ?? 0n;
    }
    return 0n;
  }

  /**
   * Get nonce from EVM
   */
  private getNonce(evm: EvmLike, address: string): bigint {
    // Check for map-based access
    if (evm.nonces) {
      return evm.nonces.get(address) ?? 0n;
    }

    // Check for method-based access
    if (evm.getNonce) {
      return evm.getNonce(address);
    }

    return 0n;
  }

  /**
   * Get code from EVM
   */
  private getCode(evm: EvmLike, address: string): Uint8Array {
    if (evm.code) {
      return evm.code.get(address) ?? new Uint8Array(0);
    }
    return new Uint8Array(0);
  }

  /**
   * Normalize address to 0x-prefixed checksummed format
   */
  private normalizeAddress(address: string): string {
    // Ensure 0x prefix
    if (!address.startsWith('0x')) {
      return '0x' + address;
    }
    return address.toLowerCase();
  }

  /**
   * Convert byte array to hex string (without 0x prefix)
   */
  private bytesToHex(bytes: Uint8Array): string {
    return Array.from(bytes)
      .map(b => b.toString(16).padStart(2, '0'))
      .join('');
  }
}
